zerver/actions/message_edit.py

```diff
@@ -1274,11 +1276,22 @@ def check_update_message(
         )
         links_for_embed |= rendering_result.links_for_preview
 
-        if message.is_stream_message() and rendering_result.has_wildcard_mention():
+        if message.is_stream_message() and rendering_result.mentions_stream_wildcard:
             stream = access_stream_by_id(user_profile, message.recipient.type_id)[0]
-            if not wildcard_mention_allowed(message.sender, stream, message.realm):
+            if not stream_wildcard_mention_allowed(message.sender, stream, message.realm):
                 raise JsonableError(
-                    _("You do not have permission to use wildcard mentions in this stream.")
+                    _("You do not have permission to use stream wildcard mentions in this stream.")
+                )
+
+        if message.is_stream_message() and rendering_result.mentions_topic_wildcard:
+            topic_participant_count = len(
+                participants_for_topic(message.realm.id, message.recipient.id, message.topic_name())
+            )
+            if not topic_wildcard_mention_allowed(
+                message.sender, topic_participant_count, message.realm
+            ):
+                raise JsonableError(
+                    _("You do not have permission to use topic wildcard mentions in this topic.")
```

Comment:
We should avoid calling `participants_for_topic`, which is an expensive function, an extra time from this code path; we call `get_recipient_info` in `do_update_message`, which is where the message-send code path gets this ... that call is just after the point of this check. So I think the fix for this issue would be to refactor the `check_update_message/do_update_message` system to put the `get_recipient_info` call in `check_update_message`, and pass it through from there, so that we can do this check using the data that fetches.

It's likely not very important -- editing messages that edit `@topic` mentions will not be super common -- but it is something worth planing to fix. This might be a refactor we should defer to 9.0 (or to a lower-priority follow-up commit late in 8.0). Maybe open an issue for this that we can set as a low-priority release goal?

Reply:
Hm, It'd be 1-2 hours of work. I'll open an issue and backlink here.

Reply:
Yeah that sounds like the right plan for this.